common/workunit/workunit.cpp

```diff
@@ -2657,85 +2660,192 @@ cost_type aggregateCost(const IConstWorkUnit * wu, const char *scope, bool exclu
+                StatsScopeId sgScopeId;
+                IPropertyTree * sgPT = & iter->query();
+                const char * sgName = sgPT->queryName();
+                if (strcmp(sgName, "node")==0)
+                    continue;
+                verifyex(sgScopeId.setScopeText(sgName));
+                MemoryBuffer compressed;
+                sgPT->getPropBin("Stats", compressed);
+                if (!compressed.length())
+                    break;
+                MemoryBuffer serialized;
+                decompressToBuffer(serialized, compressed);
+
+                unsigned version;
+                serialized.read(version);
+                byte kind;
+                serialized.read(kind);
+
+                StatsScopeId childId;
+                childId.deserialize(serialized, version);
+                int statsMinDepth = 0, statsMaxDepth = INT_MAX;
+                if (aggregatesOnly)
+                {
+                    // Only store stats for subgraph level
+                    statsMinDepth = 3;  // this is subgraph level
+                    statsMaxDepth = 3;
+                }
+                statsCollection->deserializeChild(childId, serialized, version, statsMinDepth, statsMaxDepth);
+            }
         }
     }
-    return totalCost;
 }
 
-void gatherSpillSize(const IConstWorkUnit * wu, const char *scope, stat_type & peakSizeSpill)
+void GlobalStatisticCollection::loadGlobalAggregates(IConstWorkUnit &workunit)
 {
-    WuScopeFilter filter;
-    if (!isEmptyString(scope))
-        filter.addScope(scope);
-    else
-    {
-        filter.addScope("");
-        filter.addSource("global");
-    }
-    filter.setIncludeNesting(1);
-    filter.addOutputStatistic(StSizeGraphSpill);
-    filter.addRequiredStat(StSizeGraphSpill);
-    filter.finishedFilter();
-    Owned<IConstWUScopeIterator> it = &wu->getScopeIterator(filter);
-    peakSizeSpill = 0;
-    for (it->first(); it->isValid(); )
+    class StatsCollectionAggregatesLoader : public IWuScopeVisitor
     {
-        stat_type value = 0;
-        if (it->getStat(StSizeGraphSpill, value))
+    public:
+        StatsCollectionAggregatesLoader(IStatisticCollection * _statsCollection) : statsCollection(_statsCollection) {}
+
+        virtual void noteStatistic(StatisticKind kind, unsigned __int64 value, IConstWUStatistic & extra) override
         {
-            if (value>peakSizeSpill)
-                peakSizeSpill = value;
-            it->nextSibling();
+            statsCollection->setStatistic(extra.queryScope(), kind, value);
         }
-        else
+        virtual void noteAttribute(WuAttr attr, const char * value) override { throwUnexpected(); }
+        virtual void noteHint(const char * kind, const char * value) override { throwUnexpected(); }
+        virtual void noteException(IConstWUException & exception) override { throwUnexpected(); }
+    private:
+        Linked<IStatisticCollection> statsCollection;
+    };
+
+    WuScopeFilter filter;
```

Comment:
@ghalliday You were going to come back to do this.  Any ideas on making this more efficient?

Reply:
Yes it could be made more efficient by directly walking the stats list.  If this works I wouldn't change it for this PR though - something to return to and simplify the code.